### ISSUE TITLE
kvs: add defensive checkpoints

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#4301]
Problem: currently the kvs root reference is not periodically updated while an instance runs, so a  rank 0 crash could result in significant loss of data.

The root reference is currently written out in the following circumstances
- what the start event is posted to `admin.eventlog` in rc1
- when the kvs is unloaded in rc3

Edit: note that the content cache on rank 0 must be flushed to the backing store before the checkpoint is written, otherwise a mistimed crash could result in dangling references.

**Comment by garlick** [comment]
One thought is as a first cut, we could add a `flux kvs checkpoint` subcommand which
- calls `flux_kvs_getroot()`
- makes the `content.flush` RPC on rank 0
- calls  `kvs_checkpoint_commit()`

And then add a flux cron job for doing this periodically.

What would be really nice, but possibly a bit of work, would be to add a new KVS flag that could be added to `flux_kvs_commit()` so that once the commit finishes, we are guaranteed the result is on the backing store.   Then if we could judiciously use this in places where "durability" of the commit is critical.   It would mean adding the above sequence to the KVS commit logic for the primary namespace only.  If @chu11 thinks this might be feasible, perhaps we could open a separate issue on that and let this one be closed with something simple like the cron job.

**Comment by garlick** [comment]
I opened a separate issue on the KVS flag so we can consider it more carefully there.

Meanwhile, let's say we can close this issue if we implement a mechanism for the kvs itself to periodically update the checkpoint if there has been commit activity and no checkpoint for some period of time.   This code would be useful even after a flag is added.

**Comment by chu11** [comment]
pondering this a bit yesterday. thinking about defensive checkpointing in a range, from most aggressive to least

1) after every commit
2) after every stdio commits
3) after every eventlog entry
4) after every job completes
5) every X time period (i.e. via cron)
6) every time user / sysadmin requests (i.e. checkpoint command per comments above)
7) after flux shuts down

could be missing some other options that I'm not thinking about right now.

I think 1 & 2 are nuts, and probably 3 too.  But for the system instance, I could see 4 as potential option, b/c high throughput jobs from users should be through their own instance (presumably user checkpoints will not be as often).  So the checkpointing shouldn't be nuts?

Another approach might be to support a checkpointing option within the system instance, letting admins configure for the safety level they wish, sorta like with verbosity in logs.

Nevertheless, might be worthwhile to do some experiments to see how much slower throughput is after #4348 goes in with a number of the options above.

**Comment by garlick** [comment]
Hmm, I like 5 b/c I think anything tied to jobs or more frequently is going to really slow things down.

What about having an optional kvs config like
```toml
[kvs]
# sync primary KVS namespace to backing store on specified period
# (default: no automatic sync)
sync = "30m"
```
And then just have the kvs module take care of it internally with a reactor timer.  Ideally, include some logic to only commit IFF configured period has elapsed since checkpoint was written AND the root version has incremented since then.

**Comment by chu11** [comment]
> tied to jobs or more frequently is going to really slow things down.

For 1-3 i agree, but at a system instance level presumably users will not do anything high throughput.  They will get their own sub-instance for doing that, so the checkpoint shouldn't be too often?

Or is this more of a "that's what users should be doing, but assume the worst case, which is they won't do that".

**Comment by garlick** [comment]
Between 4 and 5, I guess I could go either way.  Maybe we could allow either or both to be configured since we may need to see how this goes in production?

**Comment by chu11** [comment]
As an experiment, implemented a "after job completes" FLUX_KVS_SYNC in the job-manager

```
diff --git a/src/modules/job-manager/event.c b/src/modules/job-manager/event.c
index dd195496d..18fa86b2a 100644
--- a/src/modules/job-manager/event.c
+++ b/src/modules/job-manager/event.c
@@ -74,6 +74,7 @@ struct event_batch {
     flux_future_t *f;
     json_t *state_trans;
     zlist_t *responses; // responses deferred until batch complete
+    int flags;
 };
 
 static struct event_batch *event_batch_create (struct event *event);
@@ -129,7 +130,7 @@ static void event_batch_commit (struct event *event)
          * corresponding event in the KVS.
          */
         if (batch->txn) {
-            if (!(batch->f = flux_kvs_commit (ctx->h, NULL, 0, batch->txn)))
+            if (!(batch->f = flux_kvs_commit (ctx->h, NULL, batch->flags, batch->txn)))
                 goto error;
             if (flux_future_then (batch->f, -1., commit_continuation, batch) < 0)
                 goto error;
@@ -244,7 +245,8 @@ static int event_batch_start (struct event *event)
 
 static int event_batch_commit_event (struct event *event,
                                      struct job *job,
-                                     json_t *entry)
+                                     json_t *entry,
+                                     int flags)
 {
     char key[64];
     char *entrystr = NULL;
@@ -264,6 +266,8 @@ static int event_batch_commit_event (struct event *event,
         free (entrystr);
         return -1;
     }
+    if (flags & EVENT_SYNC)
+        event->batch->flags |= FLUX_KVS_SYNC;
     free (entrystr);
     return 0;
 }
@@ -800,7 +804,7 @@ int event_job_post_entry (struct event *event,
     if (event_job_cache (event, job, name) < 0)
         return -1;
     if (!(flags & EVENT_NO_COMMIT)
-        && event_batch_commit_event (event, job, entry) < 0)
+        && event_batch_commit_event (event, job, entry, flags | EVENT_SYNC) < 0)
         return -1;
     if (job->state != old_state) {
         double timestamp;
diff --git a/src/modules/job-manager/event.h b/src/modules/job-manager/event.h
index 2e42309b0..ee38ea370 100644
--- a/src/modules/job-manager/event.h
+++ b/src/modules/job-manager/event.h
@@ -27,6 +27,7 @@ enum job_manager_event_flags {
      *   the eventlog in the KVS.
      */
     EVENT_NO_COMMIT = 1,
+    EVENT_SYNC = 2,
 };

```

with some throughput tests on my laptop, unsurprisingly throughput was a nice chunk slower.

before
```
fluxorama|/usr/src 2>./src/test/throughput.py -n 1024
number of jobs: 1024
submit time:    2.629 s (389.5 job/s)
script runtime: 21.951s
job runtime:    21.246s
throughput:     48.2 job/s (script:  46.6 job/s)
```

after
```
number of jobs: 1024
submit time:    2.983 s (343.3 job/s)
script runtime: 25.348s
job runtime:    24.588s
throughput:     41.6 job/s (script:  40.4 job/s)
```

but when running real jobs instead of simulated ones, there was virtually no change (in this case, the after case was actually a smidge faster):

```
number of jobs: 1024
submit time:    2.969 s (344.9 job/s)
script runtime: 91.367s
job runtime:    90.654s
throughput:     11.3 job/s (script:  11.2 job/s)
```

vs

```
number of jobs: 1024
submit time:    3.168 s (323.2 job/s)
script runtime: 90.546s
job runtime:    89.628s
throughput:     11.4 job/s (script:  11.3 job/s)
```

I think the fact that the job-manager batches commits of events together is probably want makes this not as horrible as we thought it would be (i.e. I bet batches of job completions are probably being written out in 1 sync, vs. one at a time).

Granted, this was on my laptop with just 2 cores, and in docker, so take it with a grain of salt.  But I think it's worth adding as an option, something we can consider in production at a later time.

**Comment by garlick** [comment]
If we were going to add this I would suggest not creating an event flag since event posting is asynchronous, and as you pointed out, batched.  A global flag probably makes more sense to me (e.g. one set from TOML config and always goes in the `flux_kvs_commit()` call in the batch machinery.

However if the syncing is tied to the batch timer (hard wired to 10 ms) rather than job state transitions or events, then it seems like it would be much better to have a separate timer for syncing with a configurable interval.  E.g. start it on a kvs commit and when it expires send an empty commit with the sync flag.

However if we're doing _that_, it seems like we ought to have a sync timer that is triggered by _any_ kvs activity not just job activity.    So maybe do it in the KVS with the toml config suggested above?

**Comment by chu11** [comment]
> However if we're doing that, it seems like we ought to have a sync timer that is triggered by any kvs activity not just job activity. So maybe do it in the KVS with the toml config suggested above?

I was initially aiming for two different configs, one in the kvs (a `sync = "30m"` kinda config), then an alternate one in the job-manager for after each job completed.

I had forgotten about the batch commits in the job-manager, so that sort of complicated things.

In light of your comments, perhaps we'll just stick the kvs sync timer, and we can consider future options if necessary.
